File: backend/schemas/pipeline.py
"""Pydantic schemas for the pipeline-stage / DoI API [A-PS-05] [A-PS-06].

The router exposes four shapes:

- ``PipelineStateResponse`` — full GET payload, embeds ``GateStatus``.
- ``StageTransitionRequest`` — body for POST .../transition.
- ``AICouncilUpdate`` — body for PUT .../ai-council ([A-DOI-03]).
- ``WithinCutoffSet`` — body for PUT .../within-cutoff (manual setter for A2;
  A3 replaces with the computed value driven by the envelope walk).
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class GateStatus(BaseModel):
    """Snapshot of the active DoI gate vs. the project's current data state."""

    current_doi: Optional[int]
    next_doi: Optional[int]
    missing_fields: list[str] = Field(default_factory=list)
    can_advance: bool
    override_available: bool


class PipelineStateResponse(BaseModel):
    """Full pipeline state for a project, returned by GET .../pipeline."""

    project_id: str
    pipeline_stage: Optional[str]
    doi: Optional[int]
    frozen_doi: Optional[int]
    ai_council_approved: bool
    ai_council_doc_url: Optional[str]
    within_cutoff: Optional[bool]
    transitions_available: list[str] = Field(default_factory=list)
    gate_status: GateStatus


class StageTransitionRequest(BaseModel):
    """POST body for stage transitions.

    ``target_doi`` is optional — when omitted the service applies the default
    DoI for the target stage. ``override_reason`` is required when crossing a
    failed DoI gate or leaving Cancelled per [A-PS-10] / [A-BK-30].
    """

    target_stage: str
    target_doi: Optional[int] = Field(default=None, ge=0, le=5)
    override_reason: Optional[str] = None


class AICouncilUpdate(BaseModel):
    """PUT body for setting the AI Council screening flag + document link."""

    ai_council_approved: bool
    ai_council_doc_url: Optional[str] = None


class WithinCutoffSet(BaseModel):
    """PUT body for the manual within_cutoff setter (A2-only).

    A3 will replace this endpoint with the computed value coming out of the
    ranking engine's envelope walk per [A-PS-06]. ``reason`` is captured in
    the audit log so manual overrides are traceable.
    """

    within_cutoff: bool
    reason: Optional[str] = None

File: frontend/src/types/pipeline.ts
/**
 * Pipeline / DoI / intake types — mirrors backend
 * `schemas/pipeline.py` and `schemas/intake.py`.
 */

export interface PipelineGateStatus {
  current_doi: number | null;
  next_doi: number | null;
  missing_fields: string[];
  can_advance: boolean;
  override_available: boolean;
}

export interface PipelineState {
  project_id: string;
  pipeline_stage: string | null;
  doi: number | null;
  frozen_doi: number | null;
  ai_council_approved: boolean;
  ai_council_doc_url: string | null;
  within_cutoff: boolean | null;
  transitions_available: string[];
  gate_status: PipelineGateStatus;
}

export interface StageTransitionRequest {
  target_stage: string;
  target_doi?: number | null;
  override_reason?: string | null;
}

export interface IntakeProjectCreate {
  name: string;
  description: string;
  lob_id: string;
  project_type: 1 | 2 | 3;
  capex_opex?: 'capex' | 'opex';
  start_month: string; // YYYY-MM
  end_month?: string | null;
  pl_person_id?: string | null;
  is_service?: boolean;
}

export interface IntakeProjectResponse {
  id: string;
  name: string;
  pipeline_stage: string;
  doi: number;
  status: string;
  pl_person_id: string | null;
  project_type: number;
  composite_score: number | null;
  total_budget: number | null;
}
